### ISSUE TITLE
Error "An item with the same key has already been added." during query execution

### ISSUE THREAD

**Issue by panuckov-nick** [linq2db/linq2db#1313]
Prerequisites: you have object **A** with property  **Prop** in one namespace and you have object **B** that is inherited from object **A** and hiding property **Prop** by using new keyword in property notation.
```cs
public class A {
    public int Prop {get; set;}
}

public class B : A {
    public new long Prop {get; set;}
}
```
When you are using object **B** as a result of the query you getting error, in my case property ***Prop** even was not used in the query - it was just a part of used object **B**.

Reason to have such strange behavior is step by step migration from one set of business object to other that is located in different namespace.

Exception message: "An item with the same key has already been added."
Stack trace:
```
 at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at LinqToDB.Linq.Builder.ExpressionBuilder.ProcessProjection(Dictionary`2 members, Expression expression) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 2923
   at LinqToDB.Linq.Builder.SelectContext..ctor(IBuildContext parent, LambdaExpression lambda, IBuildContext[] sequences) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SelectContext.cs:line 51
   at LinqToDB.Linq.Builder.SelectBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SelectBuilder.cs:line 50
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 175
   at LinqToDB.Linq.Builder.FirstSingleBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\FirstSingleBuilder.cs:line 25
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 175
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]() in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 146
   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Query.cs:line 280
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Query.cs:line 233
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache) in C:\projects\linq2db\Source\LinqToDB\Linq\ExpressionQuery.cs:line 84
   at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression) in C:\projects\linq2db\Source\LinqToDB\Linq\ExpressionQuery.cs:line 165
   at Spotzer.AdDirection.Finisher.DataServices.CampaignService.<GetBusinessDetails>d__11.MoveNext() in C:\Repos\spotzer-ads-editor\EditorApplication\Spotzer.AdDirection.Finisher\DataServices\CampaignService.cs:line 104
```

**Comment by Corey-M** [comment]
Should be possible to fix this in [TypeAccessorT.cs][1] by adding to the filter for valid members on line 47.

Change this:

    if (memberInfo.IsFieldEx() || memberInfo.IsPropertyEx() && ((PropertyInfo)memberInfo).GetIndexParameters().Length == 0)

To this (formatting for readability only):

    if (memberInfo.IsFieldEx() || memberInfo.IsPropertyEx() &&
        ((PropertyInfo)memberInfo).GetIndexParameters().Length == 0 &&
        !_members.Any(_ => _.Name == memberInfo.Name))

This would ensure that there are no conflicting entries in the `Members` list.

The list of members returned by reflection will always have the members from the most-derived class first, then progresses up the inheritance chain with the members from the base class last.  The first member with any particular name will be the one you want, so any subsequent members with the same name should be discarded.

Alternatively it might be a good idea to fix this in [ReflectionExtensions.cs][2] by trimming duplicates from the list in `GetPublicInstanceMembersEx()` and `GetStaticMembersEx()`.  This would do the trick:

    static MemberInfo[] RemoveDuplicatesByName(this MemberInfo[] members)
    {
        return members.GroupBy(m => m.Name).Select(g => g.First()).ToArray();
    }

Since the first member of each grouping will be the most-derived version this will drop any duplications.

I can do a PR for this once I clear my other one.

[1]: https://github.com/linq2db/linq2db/blob/b182372bc62367a134670212d0ae92df3c2f9b30/Source/LinqToDB/Reflection/TypeAccessorT.cs#L47
[2]: https://github.com/linq2db/linq2db/blob/b182372bc62367a134670212d0ae92df3c2f9b30/Source/LinqToDB/Extensions/ReflectionExtensions.cs#L119

**Comment by MaceWindu** [comment]
@Corey-M PRs are allways welcome :)

**Comment by MaceWindu** [comment]
Fixed by #1336

**Comment by MaceWindu** [comment]
It's released. Can you create new issue with more details?

**Comment by giuliohome** [comment]
ok, thanks, will do soon.
